File: bin/TargetDepth/targetDepth.pl
#!/usr/bin/env perl

use strict;
use warnings;
use File::Basename;
use Parallel::ForkManager;
use Cwd qw(cwd abs_path);
use Sort::Key::Natural qw(natsort);
use Getopt::Long;
our $dirname = dirname (__FILE__);

# Check for TargetDepth binary execution
# Do not move this script from the actual directory
my $targetDepthExe = ( -x "$dirname/TargetDepth" )
? "$dirname/TargetDepth"
: die " ERROR: Cannot execute TargetDepth\n";

# Here we will check if samtools is set on Path to get total read numbers
my $samtools = `which samtools`;
chomp $samtools;

my $cat = `which cat`;
chomp $cat;

my $cut = `which cut`;
chomp $cut;

my $paste = `which paste`;
chomp $paste;

my $sed = `which sed`;
chomp $sed;

my $rm = `which rm`;
chomp $rm;

if (!$samtools) {
	print " WARNING: SAMtools was not set on PATH, which is not mandatory, but required to get total number of reads if needed\n";
}

my $input;
my $outdir;
my $outName;
my $report_counts;
my $report_coverage;
my $bed;
my $version;
my $genome;
my $verbose;
# Default cpu's is 1. Using more than 4 threads does not improve performance due to IO Disk constraints.
my $threads = 1;

# Displaying usage panel
Help () if (@ARGV<1 or !GetOptions(
	'i=s' =>\$input,
	'o=s' =>\$outdir,
	'n=s' =>\$outName,
	'c' =>\$report_counts,
	'd' =>\$report_coverage,
	'b=s' =>\$bed,
	'g=s' =>\$genome,
	't=i' =>\$threads,
	'v' =>\$version,
	'verbose' =>\$verbose,
	)
);

 if ($version) {
	# Following https://en.wikipedia.org/wiki/Software_versioning
	print " targetDepth.pl v1.0\n";
	exit;
 }

 # Now checking input bams
 if (!$input) {
	print " ERROR: No input file was provided. You can do it like this:
			option1: introduce input directory. All bam files will be analyzed
			option2: Introduce comma separated bams(BAM1,BAM2..)
			option3: Specify bam paths in a text file (one path per line)\n";
	exit;
 }
 if (!$genome) {
	print " ERROR: No genome in FASTA format was introduced\n";
	exit;
 }
 if (!-e $genome) {
	print " ERROR: $genome genome does not exist\n";
	exit;
 }
 if (!$bed) {
	print " ERROR: No BED file introduced\n";
	exit;
 }
 if (!-e $bed) {
	print " ERROR: $bed genome does not exist\n";
	exit;
 }

 #Checking BED consistency
 open (BED, "<$bed") || die " ERROR: Unable to open $bed\n";
 while (my $line=<BED>) {
	my @tmp = split (/\t/, $line);
	if (@tmp < 4) {
		print " ERROR: bed file requires a fourth field containing region information (e.g gene name)\n";
		exit;
	}
 }
 close BED;

 my $depthOption;
 if ($report_counts && $report_coverage) {
	$depthOption = "-c -d";
 }
 elsif (!$report_counts && $report_coverage) {
	$depthOption = "-d";
 }
 elsif ($report_counts && !$report_coverage) {
	$depthOption = "-c";
 }

 my $date = localtime();

 if ($verbose) {
 	print " CMD: $0 -i $input -g $genome -b $bed -o $outdir $depthOption \n\n";
 }

 # Adding multi BAM input option separated by comma
 my @bams;
 my @tmp = split ("," , $input);
 if (@tmp > 1 ) {
	print " INFO: multiple input samples recognized\n";
	foreach my $sample ( natsort @tmp) {
		print " INFO: sample $sample\n";
		push @bams, $sample;
	}
 }
 if ( -d $input ) {
	print " INFO: Analyzing all bam files from $input directory\n";
	@bams = glob "$input/*bam";
 }
 else {
	if (-B $input) {
		print " INFO: Binary Alignment Map $input\n";
	}
	elsif (-f $input) {
		print " INFO: Checking list of BAM files\n";
		open (IN, "<", $input) || die " ERROR: Cannot open $input\n";

		while (my $sample=<IN>) {
			chomp $sample;
			if (!-e $sample || !-B $sample) {
				print " ERROR: $sample is not BAM file\n";
				exit;
			}
			push @bams, $sample;
		}
		close IN;
	}
 }

 # Now do the job
 if (!-e $outdir) {
	mkdir $outdir;
 }
 my $pm = Parallel::ForkManager->new($threads);

 my $index = 0;
 my $count = 0;
 my @batch;
 foreach my $bam (natsort @bams) {

	$count++;
	$index++;
	push @batch, $bam;
	if ($count == $threads || $index == scalar @bams) {
		 foreach my $bam (@batch) {
			print " INFO: Extracting $bam read depth\n";
			my $pid = $pm -> start() and next;
			if (!-e "$bam.bai" && $samtools) {
				print " INFO: Indexing $bam\n";
				my $cmd = "$samtools index $bam\n";
				system($cmd);
			}
			my $cmd = "$targetDepthExe -i $bam $depthOption -g $genome -b $bed -o $outdir >/dev/null 2>&1";

			#my $cmd = "$targetDepthExe -i $bam $depthOption -g $genome -b $bed -o $outdir";
			if ($verbose) {
				print " INFO: CMD: $cmd\n";
			}
			system($cmd);
			print " INFO: done with $bam\n";
		   	$pm->finish;
		 }
		 $pm->wait_all_children;
		 @batch = ();
		 $count = 0;
	}
 }
 # Editing summary metrics file
 my $summary    = "$outdir/summary_metrics.log";
 my $newSummary = "$outdir/summary_metrics.tmp.log";

 if (!-e $summary || -z $summary ) {
	 print " ERROR: missing $summary file\n";

 }

 if (-e $summary ) {
	my $capture = `$cat $summary`;
	chomp $capture;
	open (IN, ">", $newSummary) || die " ERROR: Cannot open $newSummary\n";
	if ($samtools) {
		print IN "SAMPLE\tTOTAL_READS\tREADS_ON_TARGET\tREADS_CHRX\t\%ROI\tMEAN_COVERAGE\tMEAN_COUNTS\tMEAN_ISIZE\tSD_ISIZE\tMEAN_COVERAGE_X\tMEAN_COUNTS_X\n";
	}
	else {
		print IN "SAMPLE\tREADS_ON_TARGET\tREADS_CHRX\tMEAN_COVERAGE\tMEAN_COUNTS\tMEAN_ISIZE\tSD_ISIZE\tMEAN_COVERAGE_X\tMEAN_COUNTS_X\n";
	}
	my @tmpStr = split (/\n/, $capture);
	foreach my $line (@tmpStr) {

		my ($sample, $reads_on_target, $reads_X, $mean_coverage, $mean_counts,
			$mean_isize, $sd_isize, $mean_coverage_X, $mean_counts_X) = split (/\t/, $line);
		my $outsample = $sample;
		$outsample =~s/.bam//;
		if ($samtools) {
			my @samplePaths = grep ($_=~/$sample/, @bams);

       		my $total_reads = `$samtools idxstats $samplePaths[0] | awk '{i+=\$3} END {print i}'`;
			chomp $total_reads;

			my $roi = sprintf "%.2f", 100* ($reads_on_target/$total_reads);
			print IN "$outsample\t$total_reads\t$reads_on_target\t$reads_X\t$roi\t$mean_coverage\t$mean_counts\t$mean_isize\t$sd_isize\t$mean_coverage_X\t$mean_counts_X\n";
		}
		else {
			print IN "$line\n";
		}
	}
	close IN;
	#`$rm $summary`;
	rename $newSummary, $summary ;
 }


 # Merging temporal files
 my @countFiles    = glob ("$outdir/*_counts.bed");
 my @coverageFiles = glob ("$outdir/*_coverage.bed");
 my @isizeFiles = (); # Now this is deprecated
 #my @isizeFiles    = glob ("$outdir/*_isizes.bed");

 my $masterCounts = "$outdir/$outName.read.counts.bed";
 my $masterCovs   = "$outdir/$outName.per.base.coverage.bed";
 #my $masterIsizes = "$outdir/$outName.InsertSizes.bed";

 # Getting coordinates

 if($report_counts) {
 	`$cat $countFiles[0] | $cut -f 1,2,3,4,5,6 > $outdir/coords1.txt`;
 }
 if($report_coverage) {
 	`$cat $coverageFiles[0] | $cut -f 1,2,3,4,5,6 > $outdir/coords3.txt`;
 }

 # Inserting fields
 my @tmps;
 my @samp;
 my $str;
 my $header;
 if ($report_counts) {
	 foreach my $ct ( natsort @countFiles) {
		my $sample = basename($ct);
		$sample=~s/_counts.bed//;
		$sample=~s/.bam//;
		push @samp, $sample;
		`$cat $ct | $cut -f 7 > $ct.tmp`;
		push @tmps, "$ct.tmp";
	 }

	 $str = join ("\t", @samp);
	 $header = "chr\tstart\tend\texon\tgc\tmap\t$str";
	 `$paste $outdir/coords1.txt @tmps > $masterCounts `;
	 `$sed -i '1i\'"$header" $masterCounts`;

	 @tmps = ();
	 @samp = ();
	 $str = "";
	# foreach my $is (@isizeFiles) {
	#	my $sample = basename($is);
	#	$sample=~s/_isizes.bed//;
	#	$sample=~s/.bam//;
	#	push @samp, "$sample\_Mean_Isize";
	#	push @samp, "$sample\_SD_Isize";
	#	`$cat $is | $cut -f 6,7 > $is.tmp`;
	#	push @tmps, "$is.tmp";
	# }
	# $str = join ("\t", @samp);
	# $header = "chr\tstart\tend\texon\t\%GC\t$str";
	# `$paste $outdir/coords2.txt @tmps > $masterIsizes `;
	# `$sed -i '1i\'"$header" $masterIsizes`;
 }
 if ($report_coverage) {
	 @tmps = ();
	 @samp = ();
	 $str = "";
	 foreach my $cv ( natsort @coverageFiles) {
		my $sample = basename($cv);
		$sample=~s/_coverage.bed//;
		$sample=~s/.bam//;
		push @samp, $sample;
		`$cat $cv | $cut -f 7 > $cv.tmp`;
		push @tmps, "$cv.tmp";
	 }
	 $str = join ("\t", @samp);
	 $header = "chr\tstart\tend\texon\tgc\tmap\t$str";
	 `$paste $outdir/coords3.txt @tmps > $masterCovs `;
	 `$sed -i '1i\'"$header" $masterCovs`;
	 `$rm $outdir/coords3.txt`;
 	 `$rm $outdir/*_coverage.bed`;
 }

  # Deleting temporal files
 `$rm $outdir/*.tmp`;
 `$rm $outdir/coords1.txt`;
 `$rm $outdir/*_counts.bed`;
 `$rm $outdir/*_isizes.bed`;


sub Help {

 	print("
Description: targetDepth.pl
Purpose: Extract read depth metrics from a list of files
Version 0.1
Usage:  perl targetDepth.pl <OPTIONS>
options:
	-i	STRING	input BAM/s. Choose between:
			option1: introduce input directory. All bam files will be analyzed
			option2: Introduce comma-separated bams (BAM1,BAM2..)
			option3: Specify bam paths in a text file (one path per line)
	-o	STRING	Output directory. Default is '.'
	-c	BOOL	Report exon counts
	-d	BOOL	Report per base coverage
	-b	STRING	BED regions file
	-g	STRING	Genome file in FASTA format
	-t	INT	Number of threads
	-v	BOOL	Display version
\n");
exit;

}

File: bin/TargetDepth/target_depth.py
#!/usr/bin/env python3

import argparse
import os
import subprocess
import shutil
import pandas as pd
import glob
import sys
import pysam
from multiprocessing import Pool
from pathlib import Path
from collections import defaultdict

def which(program):
    return shutil.which(program)

def run_cmd(command):
    subprocess.run(command, shell=True, 
        stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)


def get_total_reads(bam_file):
    total_reads = 0
    for stat in pysam.idxstats(bam_file).splitlines():
        fields = stat.split('\t')
        total_reads += int(fields[2])
    return total_reads


def process_bam(bam_args):
    bam_file, args, depth_option = bam_args
    """ """
    # Determine the directory in which the Python script is located
    dirname = os.path.dirname(os.path.abspath(__file__))

    bam_name = os.path.basename(bam_file)

    sample_counts_file = f"{args.outdir}/{bam_name}_counts.bed"
    sample_coverage_file = f"{args.outdir}/{bam_name}_coverage.bed"
    # print(sample_counts_file, sample_coverage_file)

    # Check for TargetDepth binary execution
    targetDepthExe = os.path.join(dirname, "TargetDepth")
    if not os.path.isfile(targetDepthExe) or not os.access(targetDepthExe, os.X_OK):
        print("ERROR: Cannot execute TargetDepth")
        sys.exit(1)

    # bam_file, args = bam_args
    print(f"INFO: Processing {bam_file}")

    # Assuming targetDepthExe is a callable script
    cmd = f"{targetDepthExe} -i {bam_file} {depth_option} -g {args.genome} -b {args.bed} -o {args.outdir}"
    print(f"INFO: CMD: {cmd}")
    # if args.verbose:
    #     print(f"INFO: CMD: {cmd}")
    if not os.path.isfile(sample_counts_file) and not os.path.isfile(sample_coverage_file):
        run_cmd(cmd)
    print(f"INFO: Finished {bam_file}")


def merge_files(file_list, output_file):
    """ """
    # Extract sample names from file names
    sample_names = [os.path.basename(f).replace("_counts.bed", "")\
        .replace("_coverage.bed", "").replace(".bam", "") for f in file_list]

    # Create header with sample names
    header = "chr\tstart\tend\texon\tgc\tmap\t" + "\t".join(sample_names)
    sample_depth_dict = {}
    sample_depth = []

    for input_file in file_list:
        sample_name = os.path.basename(input_file).replace("_counts.bed", "").replace(".bam", "")

        if not sample_name in sample_depth_dict:
            sample_depth_dict[sample_name] = []

        with open (input_file) as f:
            for line in f:
                line = line.rstrip("\n")
                if line.startswith("chr\t"):
                    continue
                tmp = line.split("\t")
                depth_data = tmp[-1]
                sample_depth_dict[sample_name].append(depth_data)
                # for data in sample_depth_dict:
                #     sample_depth_dict[sample_name].append(depth_data)

        f.close()
    
    baseline_file = file_list[0]
    o = open(output_file, "w")
    o.write(header+"\n")
    idx = 0
    with open(baseline_file) as f:
        for line in f:
            line = line.rstrip("\n")
            tmp = line.split("\t")
            list_depth = []
            for sample in sample_depth_dict:
                sample_depth = sample_depth_dict[sample][idx]
                list_depth.append(sample_depth)
            o.write('\t'.join(tmp[0:6])+ "\t" + '\t'.join(list_depth)+"\n")
            idx+=1
    o.close()

    # print(file_list)


def main():
    parser = argparse.ArgumentParser(description="Extract read depth metrics from a list of files")
    parser.add_argument('-i', '--input', required=True, help="Input BAM/s. Options: directory, comma-separated bams, or file with bam paths")
    parser.add_argument('-o', '--outdir', default='.', help="Output directory")
    parser.add_argument('-c', '--report_counts', action='store_true', help="Report exon counts")
    parser.add_argument('-d', '--report_coverage', action='store_true', help="Report per base coverage")
    parser.add_argument('-b', '--bed', required=True, help="BED regions file")
    parser.add_argument('-n', '--output_name', required=True, help="Output name")
    parser.add_argument('-g', '--genome', required=True, help="Genome file in FASTA format")
    parser.add_argument('-t', '--threads', type=int, default=1, help="Number of threads")
    parser.add_argument('-v', '--version', action='store_true', help="Display version")
    parser.add_argument('--verbose', action='store_true', help="Verbose output")
    args = parser.parse_args()

    if args.version:
        print("targetDepth.py v1.0")
        exit()

    # Check BED consistency
    with open(args.bed, 'r') as bed_file:
        for line in bed_file:
            if len(line.strip().split('\t')) < 4:
                print("ERROR: bed file requires a fourth field containing region information (e.g gene name)")
                exit()

    # Determine depth option based on arguments
    depth_option = ''
    if args.report_counts and args.report_coverage:
        depth_option = '-c -d'
    elif args.report_counts:
        depth_option = '-c'
    elif args.report_coverage:
        depth_option = '-d'

    bams = []
    if os.path.isdir(args.input):
        print(f" INFO: Analyzing all bam files from {args.input} directory")
        bams = [str(bam_file) for bam_file in Path(args.input).glob("*.bam")]
    elif os.path.isfile(args.input):
        with open(args.input, 'r') as file:
            bams = [line.strip() for line in file if os.path.isfile(line.strip())]
    else:
        bams = args.input.split(',')

    if not bams:
        print(" ERROR: No valid input files found")
        exit()

    print(f"Number of threads: {args.threads}")

    # Create output directory if it doesn't exist
    os.makedirs(args.outdir, exist_ok=True)
    
    bam_arguments = [(bam_file, args, depth_option) for bam_file in bams]

    print(f" INFO: Processing {len(bams)} BAM files with {args.threads} processes")
    with Pool(processes=args.threads) as pool:
        pool.map(process_bam, bam_arguments)

    # Generating summary metrics
    summary_file = os.path.join(args.outdir, "summary_metrics.log")
    new_summary_file = os.path.join(args.outdir, "summary_metrics.tmp.log")

    samtools_path = which("samtools")
    
    if not os.path.exists(summary_file) or os.path.getsize(summary_file) == 0:
        print(f" ERROR: missing {summary_file} file")
    else:
        with open(summary_file, 'r') as rf:
            lines = rf.readlines()

        with open(new_summary_file, 'w') as sf:

            if samtools_path:
                sf.write("SAMPLE\tTOTAL_READS\tREADS_ON_TARGET\tREADS_CHRX\t%ROI\tMEAN_COVERAGE\tMEAN_COUNTS\tMEAN_ISIZE\tSD_ISIZE\tMEAN_COVERAGE_X\tMEAN_COUNTS_X\n")
            else:
                sf.write("SAMPLE\tREADS_ON_TARGET\tREADS_CHRX\tMEAN_COVERAGE\tMEAN_COUNTS\tMEAN_ISIZE\tSD_ISIZE\tMEAN_COVERAGE_X\tMEAN_COUNTS_X\n")

            for line in lines:
                data = line.strip().split('\t')
                if line.startswith("SAMPLE\t"):
                    continue
                sample = data[0]
                sample_bam = os.path.join(args.input, sample)

                total_reads = get_total_reads(sample_bam)

                reads_on_target = int(data[1])
                reads_X = int(data[2])
                mean_coverage = float(data[3])
                mean_counts = float(data[4])
                mean_isize = float(data[5]) 
                sd_isize = float(data[6])
                mean_coverage_X = float(data[7])
                mean_counts_X = float(data[8]) 

                roi = (reads_on_target / total_reads) * 100 if total_reads > 0 else 0

                sf.write(f"{sample}\t{total_reads}\t{reads_on_target}\t{reads_X}\t{roi}\t{mean_coverage}\t{mean_counts}\t{mean_isize}\t{sd_isize}\t{mean_coverage_X}\t{mean_counts_X}\n")
        rf.close()
        sf.close()
        os.remove(summary_file)
        os.rename(new_summary_file, summary_file)

    # Merging temporary files
    count_files = glob.glob(os.path.join(args.outdir, "*_counts.bed"))
    coverage_files = glob.glob(os.path.join(args.outdir, "*_coverage.bed"))

    master_counts = os.path.join(args.outdir, f"{args.output_name}.read.counts.bed")
    master_coverage = os.path.join(args.outdir, f"{args.output_name}.per.base.coverage.bed")

    # Assuming that the first file in each list contains the header
    # Modify as needed based on your file format
    if args.report_counts:
        merge_files(count_files, master_counts)
    if args.report_coverage:
        merge_files(coverage_files, master_coverage)

    # Clean up temporary files
    # for f in count_files + coverage_files:
    #     os.remove(f)

    # temp_files = glob.glob(os.path.join(args.outdir, "*.tmp"))
    # for temp_file in temp_files:
    #     os.remove(temp_file)

    # Final message or summary
    print("Processing complete. Results are available in:", args.outdir)


if __name__ == "__main__":
    main()
